segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -263,56 +211,40 @@ private Append getNextAppend() {
         }
     }
 
-    private void handleAppendResult(final Append append, Long newWriteOffset, Throwable exception, Timer elapsedTimer) {
+    private void handleAppendResult(final Append append, Long newWriteOffset, Long previousEventNumber, Throwable exception, Timer elapsedTimer) {
         boolean success = exception == null;
         try {
             boolean conditionalFailed = !success && (Exceptions.unwrap(exception) instanceof BadOffsetException);
-            long previousEventNumber;
-            synchronized (lock) {
-                previousEventNumber = latestEventNumbers.get(Pair.of(append.getSegment(), append.getWriterId()));
-                Preconditions.checkState(outstandingAppend == append,
-                        "Synchronization error in: %s while processing append: %s.",
-                        AppendProcessor.this.getClass().getName(), append);
-            }
 
+            WriterState state = this.writerStates.getOrDefault(Pair.of(append.getSegment(), append.getWriterId()), null);
             if (success) {
-                final DataAppended dataAppendedAck = new DataAppended(append.getRequestId(), append.getWriterId(), append.getEventNumber(),
-                        previousEventNumber, newWriteOffset);
-                log.trace("Sending DataAppended : {}", dataAppendedAck);
-                connection.send(dataAppendedAck);
+                Preconditions.checkState(state != null, "Synchronization error while processing append: %s. Unable to send ack.", append);
+                synchronized (state.getAckLock()) {
+                    // Acks must be sent in order. The only way to do this is by using a lock.
+                    long previousLastAcked = state.appendSuccessful(append.getEventNumber());
+                    if (previousLastAcked < append.getEventNumber()) {
```

Comment:
This `if` performs similar logic to the `max` inside of the appendSuccessful and neither works without the other. 
As these are split in two locations it makes it hard to reason about. It would be better to have all the logic either here or in one private function so that it's obvious previousLastAsked is correct.

Reply:
Actually this one does a slightly different job than `appendSuccessful`.
- `appendSuccessful` records the fact that an append with an EventNumber is complete, and updates the state. It is meant to record the last Event Number that was ack-ed.
- This piece of code deals with actually sending the Ack. I do not understand why the client needs to know both the previously acked event number and the current one, when the current one would have been enough. It also requires them to be sent in the correct order (i.e., 10-15 must be before 15-19, and not vice-versa).  If they aren't, then it will throw an exception and close the connection. 
- I chose it to have two separate `synchronized` blocks because I did not  want sync-ing for acks to interfere with updating the `lastStoredEventNumber`. Updating 2 fields is faster than sending the ack over the wire so I didn't want to block any new appends on that completing.